omp/matrix/csr_kernels.cpp

```diff
@@ -127,6 +129,199 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+                        size_type row)
+{
+    auto a_row_ptrs = a->get_const_row_ptrs();
+    auto a_col_idxs = a->get_const_col_idxs();
+    auto b_row_ptrs = b->get_const_row_ptrs();
+    auto b_col_idxs = b->get_const_col_idxs();
+    for (size_type a_nz = a_row_ptrs[row];
+         a_nz < size_type(a_row_ptrs[row + 1]); ++a_nz) {
+        auto a_col = a_col_idxs[a_nz];
+        auto b_row = a_col;
+        cols.insert(b_col_idxs + b_row_ptrs[b_row],
+                    b_col_idxs + b_row_ptrs[b_row + 1]);
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+void spgemm_accumulate_row(std::unordered_map<IndexType, ValueType> &cols,
+                           const matrix::Csr<ValueType, IndexType> *c,
+                           ValueType scale, size_type row)
+{
+    auto row_ptrs = c->get_const_row_ptrs();
+    auto col_idxs = c->get_const_col_idxs();
+    auto vals = c->get_const_values();
+    for (size_type c_nz = row_ptrs[row]; c_nz < size_type(row_ptrs[row + 1]);
+         ++c_nz) {
+        auto c_col = col_idxs[c_nz];
+        auto c_val = vals[c_nz];
+        cols[c_col] += scale * c_val;
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+void spgemm_accumulate_row2(std::unordered_map<IndexType, ValueType> &cols,
+                            const matrix::Csr<ValueType, IndexType> *a,
+                            const matrix::Csr<ValueType, IndexType> *b,
+                            ValueType scale, size_type row)
+{
+    auto a_row_ptrs = a->get_const_row_ptrs();
+    auto a_col_idxs = a->get_const_col_idxs();
+    auto a_vals = a->get_const_values();
+    auto b_row_ptrs = b->get_const_row_ptrs();
+    auto b_col_idxs = b->get_const_col_idxs();
+    auto b_vals = b->get_const_values();
+    for (size_type a_nz = a_row_ptrs[row];
+         a_nz < size_type(a_row_ptrs[row + 1]); ++a_nz) {
+        auto a_col = a_col_idxs[a_nz];
+        auto a_val = a_vals[a_nz];
+        auto b_row = a_col;
+        for (size_type b_nz = b_row_ptrs[b_row];
+             b_nz < size_type(b_row_ptrs[b_row + 1]); ++b_nz) {
+            auto b_col = b_col_idxs[b_nz];
+            auto b_val = b_vals[b_nz];
+            cols[b_col] += scale * a_val * b_val;
+        }
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+void spgemm(std::shared_ptr<const OmpExecutor> exec,
+            const matrix::Csr<ValueType, IndexType> *a,
+            const matrix::Csr<ValueType, IndexType> *b,
+            const matrix::Csr<ValueType, IndexType> *c,
+            Array<IndexType> &c_row_ptrs_array,
+            Array<IndexType> &c_col_idxs_array, Array<ValueType> &c_vals_array)
+{
+    auto rows = a->get_size()[0];
+
+    // first sweep: count nnz for each row
+    c_row_ptrs_array.resize_and_reset(rows + 1);
+    auto c_row_ptrs = c_row_ptrs_array.get_data();
+
+    std::unordered_set<IndexType> local_col_idxs;
+#pragma omp parallel for schedule(dynamic, 256) firstprivate(local_col_idxs)
```

Comment:
Is any reason to set 256 as the parameter?

Reply:
This is basically a very crude attempt at limiting the overhead of dynamic scheduling. I'm open for suggestions.

Reply:
Since the default `schedule` is implementation-defined and thus could also be `static` (which would totally make sense in many loops, e.g., dense linear algebra), as a rule of thumb I put no schedule statement in loops whose iterations contain equal amounts of work. In any unbalanced loops, I make the schedule dynamic with sufficiently large block size so we don't get significant overhead from scheduling. Should I remove the schedule statement for now, and consider it again when we look at fine-tuning the OpenMP kernels?

Reply:
You can leave them as they are, but please add a comment to state that it the parameter is not necessarily optimized / that further optimizations are required.

Reply:
I decided to remove them after all, as you are probably correct, the OpenMP defaults should be somewhat sane.